### PR TITLE
Validate slice timing array with RepetitionTime

### LINKED ISSUE #302: Validate that each value in the SliceTiming array is smaller than RepetitionTime

### PR DESCRIPTION
**Pull request by suyashdb**
Closes #302.

### CODE REVIEW

**Comment by suyashdb** [comment]
@chrisfilo: Ready for review.

File: utils/issues/list.js [hunk 1]
```diff
@@ -326,5 +326,10 @@ module.exports = {
         key: 'SESSION_LABEL_IN_FILENAME_DOESNOT_MATCH_DIRECTORY',
         severity: 'error',
         reason: 'Session label in the filename doesn\'t match with the path of the file. File seems to be saved in incorrect session directory.'
+    },
+    66: {
+        key: 'SLICETIMING_VALUES_GREATOR_THAN_REPETITION_TIME',
+        severity: 'error',
+        reason: 'SliceTiming value/s in json contains invalid value as it is greator than RepetitionTime.'
```
Comment (chrisgorgo) [inline utils/issues/list.js]: Put the names of the fields in quotes (consistent with other error descriptions) and don't mention the JSON file.

File: tests/json.spec.js [hunk 2]
```diff
@@ -25,4 +25,12 @@ describe('JSON', function(){
 		});
 	});
 
+	it('sidecar SliceTiming shouldnot be greator than RepetitionTime', function(){
```
Comment (chrisgorgo) [inline tests/json.spec.js]: should not

File: tests/json.spec.js [hunk 3]
```diff
@@ -25,4 +25,12 @@ describe('JSON', function(){
...
+		var jsonObjInval = '{"RepetitionTime": 1.2, "echo_time": 0.005, "flip_angle": 90, "TaskName": "Rest", "SliceTiming":[0.0, 1.7759, 0.0517, 0.8276, 0.1034, 1.8793, 0.1552]}';
+		validate.JSON(file, jsonObjInval, function (issues) {
+			assert(issues.length === 1 && issues);
```
Comment (chrisgorgo) [inline tests/json.spec.js]: Please verify that the expected error was raised (via the error code) - not just any issue.

File: validators/json.js [hunk 4]
```diff
@@ -60,6 +60,22 @@ function checkUnits (file, sidecar) {
             code: 2
         }));
     }
+    if (sidecar.hasOwnProperty('RepetitionTime') && sidecar.hasOwnProperty('SliceTiming')) {
```
Comment (chrisgorgo) [inline validators/json.js]: This needs to be done on the level of merged dictionary not individual JSON file. RepetitionTime and SliceTiming could be defined in different files.

File: validators/json.js [hunk 5]
```diff
@@ -60,6 +60,22 @@ function checkUnits (file, sidecar) {
...
+        var SliceTimingArray = sidecar["SliceTiming"];
+        var invalid_timesArray = [];
+        SliceTimingArray.filter(function(time){
+            if (time > sidecar['RepetitionTime']){
+                invalid_timesArray.push(time);
+            }
+        });
```
Comment (chrisgorgo) [inline validators/json.js]: Why not do a for loop that goes over all values in SliceTiming and checks each against RepetitionTime. The look could break at the first incorrect value.

File: utils/issues/list.js (hunk 1 above)
Comment (chrisgorgo) [reply utils/issues/list.js]: The error should also mention that SliceTiming values should be in seconds not milliseconds (common mistake).

**Review by chrisgorgo — CHANGES_REQUESTED** [review]

File: tests/nii.spec.js [hunk 6]
```diff
@@ -135,4 +135,37 @@ describe('NIFTI', function(){
...
+        var file_new = {
+            name: 'sub-15_task-mixedeventrelatedprobe_run-01_bold.nii.gz',
+            relativePath: '/sub-15/func/sub-15_task-mixedeventrelatedprobe_run-01_bold.nii.gz'
+        };
+        var events = [
+            '/sub-15/func/sub-14_task-mixedeventrelatedprobe_run-01_events.tsv',
+            '/sub-15/run-01_events.tsv'
+        ];
+        validate.NIFTI(null, file_new, jsonContentsDict_new, {}, [], events, function (issues) {
+            assert(issues[2].code === 66);
```
Comment (chrisgorgo) [inline tests/nii.spec.js]: could you also add assertions for number of errors and warnings?

**Review by chrisgorgo — COMMENTED** [review]
Thanks! Much better!
Code climat seems to be unhappy about nested ifs https://codeclimate.com/github/INCF/bids-validator/pull/352#issues. Here are some tips how to improve this. https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting

File: tests/nii.spec.js (hunk 6 above)
Comment (suyashdb) [reply tests/nii.spec.js]: Sure will do. 
Yes i am trying to resolve the code climate errors. :)

**Comment by suyashdb** [comment]
@chrisfilo 
CircelCi and Code Climate both looks happy this time!

**Review by chrisgorgo — APPROVED** [review]

**Pull request merged**